File: shared/decomposition.py
SWITCHES = (
    "commitment",
    "fingerprint",
    "p1",
    "p2",
    "p3",
    "p4",
    "order1",
    "product1",
    "p5",
    "order2",
    "product2",
    "p6",
)

REQUIRES = {"p5": "order1", "p6": "order2"}

BINDING = ("commitment", "fingerprint")

GADGETS = ("p1", "p2", "p3", "p4")


class ConfigurationError(ValueError):
    pass


def validate(switches) -> tuple[str, ...]:
    enabled = set(switches)
    unknown = enabled - set(SWITCHES)
    if unknown:
        raise ConfigurationError(f"unknown switches {sorted(unknown)}; known are {list(SWITCHES)}")
    for switch, required in REQUIRES.items():
        if switch in enabled and required not in enabled:
            raise ConfigurationError(
                f"{switch} is an adjacency check on a claimed ordering, so it needs {required}; "
                "without it the check is the unsoundness the audit was built around"
            )
    return tuple(switch for switch in SWITCHES if switch in enabled)


def command_flag(switches) -> str:
    return "--enable=" + ",".join(validate(switches))


def parse(text: str) -> tuple[str, ...]:
    stripped = text.strip()
    if stripped in ("", "none"):
        return ()
    return validate(part.strip() for part in stripped.split(",") if part.strip())


_FLOOR: dict[str, tuple[str, ...]] = {
    "bare": (),
    "commitment_only": ("commitment",),
    "fingerprint_only": ("fingerprint",),
    "baseline": ("commitment", "fingerprint"),
}

_ISOLATED: dict[str, tuple[str, ...]] = {
    "p1_completeness": ("p1",),
    "p2_currentness": ("p2",),
    "p3_compliance": ("p3",),
    "p4_load_ratio": ("p4",),
    "order1_sortedness": ("order1",),
    "product1_permutation": ("product1",),
    "order2_sortedness": ("order2",),
    "product2_permutation": ("product2",),
    "sigma1_machinery": ("order1", "product1"),
    "sigma2_machinery": ("order2", "product2"),
    "p5_no_overlap": ("order1", "p5"),
    "p6_uniqueness": ("order2", "p6"),
    "both_sigma_machinery": ("order1", "product1", "order2", "product2"),
}

LEAVE_ONE_OUT = {switch: f"full_without_{switch}" for switch in SWITCHES}


CONFIGURATIONS: dict[str, tuple[str, ...]] = dict(_FLOOR)
for _name, _extra in _ISOLATED.items():
    CONFIGURATIONS[_name] = validate(set(_FLOOR["baseline"]) | set(_extra))
for _switch in SWITCHES:
    _remaining = set(SWITCHES) - {_switch}
    for _assertion, _ordering in REQUIRES.items():
        if _ordering not in _remaining:
            _remaining.discard(_assertion)
    CONFIGURATIONS[LEAVE_ONE_OUT[_switch]] = validate(_remaining)
CONFIGURATIONS["full"] = SWITCHES

TIMED = (
    "bare",
    "baseline",
    "p1_completeness",
    "p2_currentness",
    "p3_compliance",
    "p4_load_ratio",
    "sigma1_machinery",
    "sigma2_machinery",
    "both_sigma_machinery",
    "full",
)

File: dataset/src/decomposition.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Enabled {

    pub commitment: bool,

    pub fingerprint: bool,

    pub p1: bool,

    pub p2: bool,

    pub p3: bool,

    pub p4: bool,

    pub order1: bool,

    pub product1: bool,

    pub p5: bool,

    pub order2: bool,

    pub product2: bool,

    pub p6: bool,
}

impl Default for Enabled {
    fn default() -> Self {
        Self::all()
    }
}

impl Enabled {
    pub const NAMES: [&'static str; 12] = [
        "commitment",
        "fingerprint",
        "p1",
        "p2",
        "p3",
        "p4",
        "order1",
        "product1",
        "p5",
        "order2",
        "product2",
        "p6",
    ];

    pub fn all() -> Self {
        Self {
            commitment: true,
            fingerprint: true,
            p1: true,
            p2: true,
            p3: true,
            p4: true,
            order1: true,
            product1: true,
            p5: true,
            order2: true,
            product2: true,
            p6: true,
        }
    }

    pub fn none() -> Self {
        Self {
            commitment: false,
            fingerprint: false,
            p1: false,
            p2: false,
            p3: false,
            p4: false,
            order1: false,
            product1: false,
            p5: false,
            order2: false,
            product2: false,
            p6: false,
        }
    }

    pub fn sigma1(&self) -> bool {
        self.order1 || self.product1
    }

    pub fn sigma2(&self) -> bool {
        self.order2 || self.product2
    }

    pub fn any_grand_product(&self) -> bool {
        self.product1 || self.product2
    }

    pub fn needs_record_fingerprint(&self) -> bool {
        self.fingerprint || self.any_grand_product()
    }

    fn switch(&mut self, name: &str) -> Result<(), String> {
        match name {
            "commitment" => self.commitment = true,
            "fingerprint" => self.fingerprint = true,
            "p1" => self.p1 = true,
            "p2" => self.p2 = true,
            "p3" => self.p3 = true,
            "p4" => self.p4 = true,
            "order1" => self.order1 = true,
            "product1" => self.product1 = true,
            "p5" => self.p5 = true,
            "order2" => self.order2 = true,
            "product2" => self.product2 = true,
            "p6" => self.p6 = true,
            other => {
                return Err(format!(
                    "unknown switch {other:?}; known are {}",
                    Self::NAMES.join(", ")
                ))
            }
        }
        Ok(())
    }

    pub fn parse(text: &str) -> Result<Self, String> {
        let trimmed = text.trim();
        if trimmed.is_empty() || trimmed == "none" {
            return Ok(Self::none());
        }
        let mut enabled = Self::none();
        for name in trimmed.split(',').map(str::trim).filter(|part| !part.is_empty()) {
            enabled.switch(name)?;
        }
        enabled.validate()?;
        Ok(enabled)
    }

    pub fn from_arguments<'a>(
        arguments: impl IntoIterator<Item = &'a String>,
    ) -> Result<Self, String> {
        let mut enabled = Self::all();
        for argument in arguments {
            if let Some(value) = argument.strip_prefix("--enable=") {
                enabled = Self::parse(value)?;
            }
        }
        Ok(enabled)
    }

    pub fn validate(&self) -> Result<(), String> {
        for (switch, name, required, required_name) in [
            (self.p5, "p5", self.order1, "order1"),
            (self.p6, "p6", self.order2, "order2"),
        ] {
            if switch && !required {
                return Err(format!(
                    "{name} is an adjacency check on a claimed ordering, so it needs \
                     {required_name}; without it the check is the unsoundness the audit was built \
                     around"
                ));
            }
        }
        Ok(())
    }

    pub fn to_list(&self) -> String {
        let present = [
            self.commitment,
            self.fingerprint,
            self.p1,
            self.p2,
            self.p3,
            self.p4,
            self.order1,
            self.product1,
            self.p5,
            self.order2,
            self.product2,
            self.p6,
        ];
        let names: Vec<&str> = Self::NAMES
            .iter()
            .zip(present)
            .filter(|(_, on)| *on)
            .map(|(name, _)| *name)
            .collect();
        names.join(",")
    }
}
